### PR TITLE
refactor: re-structure the types package

### PR DESCRIPTION
**Pull request by bd82**

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,21 +1,16 @@
 version: 2.1 # use CircleCI 2.1
```
Comment (bd82) [inline .circleci/config.yml:1]: We can remove the changes I made in this file.

1. They should be done in a different PR (with node14/16 added as well).
2. I think I only did those because I thought (mistakenly) there was some issue with old node10

File: package.json [hunk 2]
```diff
@@ -78,6 +77,7 @@
 		"optional-require": "1.0.3"
 	},
 	"devDependencies": {
+		"@sap-devx/app-studio-toolkit-types": "file:./types",
```
Comment (bd82) [inline package.json:80]: This is an indicator we need to transform this repo into a mono-repo.
- **Separate** topic of tech debt.

File: sample-action-client/package.json [hunk 3]
```diff
@@ -72,10 +72,8 @@
 		"onCommand:contrib.action.open.settings",
 		"onCommand:contrib.action.open.file"
 	],
-	"dependencies": {
-		"@sap-devx/app-studio-toolkit-types": "0.0.7"
-	},
 	"devDependencies": {
+		"@sap-devx/app-studio-toolkit-types": "0.0.11",
```
Comment (bd82) [inline sample-action-client/package.json:76]: 👍

File: src/actions/actionsFactory.ts [hunk 4]
```diff
@@ -1,32 +1,32 @@
 import { Uri } from "vscode";
-import { IAction, ICommandAction, ActionType, IFileAction, ActionJsonKey, ISnippetAction } from './interfaces';
+import { IAction, ICommandAction, IFileAction, ISnippetAction, ActionType } from "../../types/api";
```
Comment (bd82) [inline src/actions/actionsFactory.ts:2]: This import should use the types package name, not a **relative** path.

File: src/actions/performer.ts [hunk 5]
```diff
@@ -16,31 +11,31 @@ export async function _performAction(action: IAction): Promise<any> {
       { action }
     );
     switch (action.actionType) {
-      case ActionType.Command: {
+      case "COMMAND": {
```
Comment (bd82) [inline src/actions/performer.ts:14]: Because we removed the enums we now have "magic" string values.
We should define/declare all of these **once** in some **constants** helper file
to enable easy future refactoring/changes.

File: src/actions/actionsFactory.ts [hunk 6]
```diff
@@ -1,32 +1,32 @@
...
 import { CommandAction, FileAction, SnippetAction } from './impl';
 
 const getNameProp = (fromSettings: boolean): string => {
```
Comment (bd82) [inline src/actions/actionsFactory.ts:5]: lots of "magic strings" in this file

File: src/actions/client.ts [hunk 7]
```diff
@@ -1,5 +1,5 @@
 import { workspace, ConfigurationTarget } from 'vscode';
-import { IAction } from './interfaces';
+import { IAction } from "../../types/api";
```
Comment (bd82) [inline src/actions/client.ts:2]: avoid relative imports from outside this package.

File: src/actions/controller.ts [hunk 8]
```diff
@@ -1,6 +1,6 @@
 import { extensions } from "vscode";
 import { getLogger } from "../logger/logger";
-import { IAction } from "./interfaces";
+import { IAction } from "../../types/api";
```
Comment (bd82) [inline src/actions/controller.ts:3]: avoid relative imports from outside this package.

File: src/actions/impl.ts [hunk 9]
```diff
@@ -1,6 +1,6 @@
 import { Uri } from "vscode";
-import { ActionType, IAction, ICommandAction, IExecuteAction, IFileAction, ISnippetAction,
-    CommandActionParams, ExecuteActionParams, SnippetActionParams, FileActionParams } from "./interfaces";
+import { IAction, ICommandAction, IExecuteAction, IFileAction, ISnippetAction,
```
Comment (bd82) [inline src/actions/impl.ts:2]: avoid relative imports from outside this package.

File: src/actions/impl.ts [hunk 10]
```diff
@@ -14,7 +14,7 @@ export class ExecuteAction extends Action implements IExecuteAction {
 
     constructor() {
         super();
-        this.actionType = ActionType.Execute;
+        this.actionType = "EXECUTE";
```
Comment (bd82) [inline src/actions/impl.ts:17]: many magical strings in this file

File: src/actions/impl.ts [hunk 11]
```diff
@@ -36,22 +36,24 @@ export class SnippetAction extends Action implements ISnippetAction {
...
         this.snippetName = "";
-        this.context = "";
+        this.context = {};
+        this.isNonInteractive = false;
     }
 }
 
 export class FileAction extends Action implements IFileAction {
-    uri: FileActionParams;
+    uri: IFileAction["uri"];
```
Comment (bd82) [inline src/actions/impl.ts:52]: 👍

File: src/actions/performer.ts [hunk 12]
```diff
@@ -1,12 +1,7 @@
 import { commands, ViewColumn } from 'vscode';
 import { get } from 'lodash';
 import { getLogger } from '../logger/logger';
-import {
-  ActionJsonKey, ActionType,
-  IAction, ICommandAction,
-  IExecuteAction, IFileAction,
-  ISnippetAction
-} from './interfaces';
+import {  IAction, ICommandAction, IExecuteAction, IFileAction, ISnippetAction, } from "../../types/api";
```
Comment (bd82) [inline src/actions/performer.ts:4]: avoid relative imports from outside this package.

File: src/api.ts [hunk 13]
```diff
@@ -1,13 +1,12 @@
 import { extensions } from 'vscode';
+import { BasToolkit } from "../types/api";
```
Comment (bd82) [inline src/api.ts:2]: avoid relative imports from outside this package.

File: tests/actionsFactory.spec.ts [hunk 14]
```diff
@@ -52,16 +50,16 @@ describe("actionsFactory test", () => {
 
         it("fails without name", () => {
             const actionJson = {
-                [ActionJsonKey.ActionType]: ActionType.Command
+                actionType: "COMMAND"
             };
-            expect(() => ActionsFactory.createAction(actionJson)).to.throw(`${ActionJsonKey.CommandName} is missing for actionType=${ActionType.Command}`);
+            expect(() => ActionsFactory.createAction(actionJson)).to.throw(`commandName is missing for actionType = COMMAND`);
```
Comment (bd82) [inline tests/actionsFactory.spec.ts:55]: generally for these kinds of assertions:
- Consider using some constants to make future refactoring easier.
- Consider using less precise error message matching using a regular expression to reduce maintenance overhead.
  - https://www.chaijs.com/api/bdd/#method_throw

File: tests/actionsFactory.spec.ts [hunk 15]
```diff
@@ -129,33 +139,34 @@ describe("actionsFactory test", () => {
...
+            expect(() => ActionsFactory.createAction(actionJson)).to.throw(`actionType is missing`);
         });
 
         it("when no unsupported action type used", () => {
             const actionJson = {
-                [ActionJsonKey.ActionType]: "Unsupported"
+                actionType: "Unsupported"
             };
-            expect(() => ActionsFactory.createAction(actionJson)).to.throw(`Action with ${ActionJsonKey.ActionType}=Unsupported could not be created from json file`);
+            expect(() => ActionsFactory.createAction(actionJson)).to.throw(`Action with actionType = Unsupported could not be created from json file`);
```
Comment (bd82) [inline tests/actionsFactory.spec.ts:161]: I think the value of the actionType in the error message should be quoted for increased readability.

File: tests/api.spec.ts [hunk 16]
```diff
@@ -3,7 +3,7 @@ import { expect, use } from "chai";
 import * as chaiAsPromised from "chai-as-promised";
 import * as _ from "lodash";
 import { SinonSandbox, SinonMock, createSandbox } from "sinon";
-import { IAction, ActionType, ICommandAction } from "../src/actions/interfaces";
+import { IAction, ICommandAction } from "../types/api";
```
Comment (bd82) [inline tests/api.spec.ts:6]: avoid relative imports of a different sub-package.

File: tests/actionsFactory.spec.ts [hunk 17]
```diff
@@ -8,11 +8,9 @@ const testVscode = {
 
 mockVscode(testVscode, "src/actions/actionsFactory.ts");
 mockVscode(testVscode, "src/actions/impl.ts");
-mockVscode(testVscode, "src/actions/interfaces.ts");
 import { ActionsFactory } from "../src/actions/actionsFactory";
 import { CommandAction, ExecuteAction, FileAction, SnippetAction } from "../src/actions/impl";
-import { ActionType, ActionJsonKey } from '../src/actions/interfaces';
-
+import { window } from "vscode";
```
Comment (bd82) [inline tests/actionsFactory.spec.ts:13]: import order (3rd party before relative)

File: tests/actionsFactory.spec.ts [hunk 18]
```diff
@@ -28,9 +26,9 @@ describe("actionsFactory test", () => {
     describe("create command action", () => {
         it("suceeds with params", () => {
             const actionJson = {
-                [ActionJsonKey.ActionType]: ActionType.Command,
-                [ActionJsonKey.CommandName]: "myCommand",
-                [ActionJsonKey.CommandParams]: ["param1", "param2"]
+                actionType: "COMMAND",
```
Comment (bd82) [inline tests/actionsFactory.spec.ts:29]: Please evaluate the usefulness of declaring the type of this const
```
const actionJson: ICommadAction ...
```

Comment (bd82) [reply tests/actionsFactory.spec.ts:29]: In general for **all** tests.

e.g maintainability? refactoring (new property added) ect...

File: tests/client.spec.ts [hunk 19]
```diff
@@ -14,7 +14,7 @@ mockVscode(testVscode, "src/actions/client.ts");
 mockVscode(testVscode, "src/actions/performer.ts");
 import { performAction } from "../src/actions/client";
 import * as performer from '../src/actions/performer';
-import { ActionType } from "../src/api";
+import { ICommandAction } from "../types/api";
```
Comment (bd82) [inline tests/client.spec.ts:17]: avoid relative imports from **other** sub packages.

File: tests/client.spec.ts [hunk 20]
```diff
@@ -28,7 +28,7 @@ describe("client test", () => {
     };
 
     const myAction = {
-        actionType: ActionType.Command,
+        actionType: "COMMAND",
```
Comment (bd82) [inline tests/client.spec.ts:31]: declare the type of ICommandAction on the const than you won't need these many casts.

Comment (bd82) [reply tests/client.spec.ts:31]: Does this object literal fully implement ICommandAction?

File: tests/client.spec.ts [hunk 21]
```diff
@@ -55,30 +55,30 @@ describe("client test", () => {
     describe("perform action", () => {
         it("performs the action without schedule", async () => {
             performerMock.expects("_performAction").withExactArgs(myAction);
-            await performAction(myAction);
+            await performAction(myAction as ICommandAction);
```
Comment (bd82) [inline tests/client.spec.ts:58]: likely un-needed cast

File: tests/client.spec.ts [hunk 22]
```diff
@@ -55,30 +55,30 @@ describe("client test", () => {
...
         });
 
         it("schedules the action with schedule (existing action list, update successful)", async () => {
             const actions: any[] = [myAction, myAction];
             workspaceMock.expects("getConfiguration").returns(config);
             configMock.expects("get").withExactArgs("actions", []).returns(actions);
             configMock.expects("update").withExactArgs("actions", [myAction, myAction, myAction], 2).resolves();
-            await performAction(myAction, { schedule: true}); 
+            await performAction(myAction as ICommandAction, { schedule: true});
```
Comment (bd82) [inline tests/client.spec.ts:66]: likely un-needed cast

File: tests/client.spec.ts [hunk 23]
```diff
@@ -55,30 +55,30 @@ describe("client test", () => {
...
-            await performAction(myAction, { schedule: true}); 
+            await performAction(myAction as ICommandAction, { schedule: true}); 
         });
 
         it("schedules the action with schedule (empy action list, update successful)", async () => {
             workspaceMock.expects("getConfiguration").returns(config);
             configMock.expects("get").withExactArgs("actions", []).returns([]);
             configMock.expects("update").withExactArgs("actions", [myAction], 2).resolves();
-            await performAction(myAction, { schedule: true});
+            await performAction(myAction as ICommandAction, { schedule: true});
```
Comment (bd82) [inline tests/client.spec.ts:73]: likely un-needed cast

File: tests/client.spec.ts [hunk 24]
```diff
@@ -55,30 +55,30 @@ describe("client test", () => {
...
         });
 
         it("schedules the action with schedule (existing action list, update rejected)", async () => {
             const actions: any[] = [myAction, myAction];
             workspaceMock.expects("getConfiguration").returns(config);
             configMock.expects("get").withExactArgs("actions", []).returns(actions);
             configMock.expects("update").withExactArgs("actions", [myAction, myAction, myAction], 2).rejects("Reasons!");
-            await performAction(myAction, { schedule: true});
+            await performAction(myAction as ICommandAction, { schedule: true});
```
Comment (bd82) [inline tests/client.spec.ts:81]: likely un-needed cast

File: tests/performer.spec.ts [hunk 25]
```diff
@@ -1,6 +1,7 @@
 import { expect } from "chai";
 import { SinonSandbox, SinonMock, createSandbox } from "sinon";
-import { ActionJsonKey, ActionType, IAction } from "../src/api";
+import { IAction, ICommandAction, IExecuteAction, IFileAction, ISnippetAction } from "../types/api";
```
Comment (bd82) [inline tests/performer.spec.ts:3]: avoid relative import from other sub-package.

File: tests/performer.spec.ts [hunk 26]
```diff
@@ -67,79 +69,71 @@ describe("performer test", () => {
...
-        });
     });
 
     describe("fileAction", () => {
         it("is fulfilled if executeCommand is fulfilled", async () => {
-            const fileAction = {
-                actionType: ActionType.File,
-                uri: 'file:///home/user/projects/myproj/sourcefile.js'
+            const fileJson = {
+                "actionType": "FILE",
```
Comment (bd82) [inline tests/performer.spec.ts:87]: quotes not needed around keys in JS (only in JSON)

File: tests/performer.spec.ts [hunk 27]
```diff
@@ -67,79 +69,71 @@ describe("performer test", () => {
...
             // check that no error is thrown
             await _performAction(fileAction);
         });
 
         it("is rejected if executeCommand rejects", async () => {
-            const fileAction = {
-                actionType: ActionType.File,
-                uri: 'file:///home/user/projects/myproj/sourcefile.js'
+            const fileJson = {
+                "actionType": "FILE",
```
Comment (bd82) [inline tests/performer.spec.ts:98]: quotes around keys not needed

File: types/package.json [hunk 28]
```diff
@@ -1,25 +1,25 @@
 {
   "name": "@sap-devx/app-studio-toolkit-types",
-  "publisher": "SAPOSS",
-  "version": "0.0.11",
+  "version": "1.0.0",
   "author": "SAP SE",
   "license": "Apache-2.0",
-  "description": "An npm package that includes Business Application Studio declarations",
+  "description": "The SAP Business Application Studio VSCode Ext type signatures",
```
Comment (bd82) [inline types/package.json:6]: I think I wrote this description, but maybe we need a better one from technical write?

File: types/api.d.ts [hunk 29]
```diff
@@ -0,0 +1,49 @@
+import {Uri} from 'vscode';
+
+declare interface BasToolkit {
+    getExtensionAPI: <T>(extensionId: string) => Promise<T>;
+    getAction: (actionId: string) => IAction | undefined;
+    getParameter: (parameterName: string) => Promise<string | undefined>;
+
+    actions: {
+        performAction: (action: IAction, options?: any) => Thenable<void>;
```
Comment (bd82) [inline types/api.d.ts:9]: 1. actions always return `Thenable<void>` ?
2. could it be `Thenable<any>` or `Thenable<T>`?

File: types/api.d.ts [hunk 30]
```diff
@@ -0,0 +1,49 @@
...
+        ExecuteAction: { new(): IExecuteAction };
```
Comment (bd82) [inline types/api.d.ts:10]: add @deprecated annotation in comment on these classes as we intend to remove them.
Maybe even some deprecation related comment text information.

File: types/api.d.ts [hunk 31]
```diff
@@ -0,0 +1,49 @@
...
+
+declare type ActionType = "EXECUTE" | "COMMAND" | "TASK" | "FILE" | "SNIPPET"
+
+declare type CommandActionParams = any[];
+declare type ExecuteActionParams = any[];
+declare type SnippetActionParams = Record<string,any>
+
+interface IAction {
+    id?: string;
+    actionType: ActionType | undefined;
```
Comment (bd82) [inline types/api.d.ts:27]: can this ever be `undefnied`?

File: types/api.d.ts [hunk 32]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+interface IExecuteAction extends IAction {
+    executeAction: (params?: ExecuteActionParams) => Thenable<any>;
```
Comment (bd82) [inline types/api.d.ts:31]: I think we discussed overriding the `actionType` in each sub-interface with a more specific value

File: types/api.d.ts [hunk 33]
```diff
@@ -0,0 +1,49 @@
+import {Uri} from 'vscode';
+
+declare interface BasToolkit {
```
Comment (bd82) [inline types/api.d.ts:3]: I think adding in-depth documentation to everything here may be part of the scope
of this task as the goal of exporting the types "properly" from technical perspective
exists to serve the higher goal of having good type signatures that are easy to use.

File: tests/actionsFactory.spec.ts (hunk 18 above)
Comment (NetaMolcho) [reply tests/actionsFactory.spec.ts:29]: actionJson is a json file, not an ICommandAction object

File: tests/client.spec.ts (hunk 20 above)
Comment (NetaMolcho) [reply tests/client.spec.ts:31]: yes, the mandatory fields in ICommandAction are actionType and name

File: sample-action-client/package.json (hunk 3 above)
Comment (bd82) [reply sample-action-client/package.json:76]: I think you need to use the following form while the project is not yet a mono-repo
- `		"@sap-devx/app-studio-toolkit-types": "file:./types",`

File: src/actions/actionsFactory.ts [hunk 34]
```diff
@@ -1,36 +1,38 @@
...
-    return fromSettings ? "name" : ActionJsonKey.CommandName;
+    return fromSettings ? "name" : "commandName";
 };
 const getParamsProp = (fromSettings: boolean): string => {
-    return fromSettings ? "params" : ActionJsonKey.CommandParams;
+    return fromSettings ? "params" : "commandParams";
 };
 
 export class ActionsFactory {
```
Comment (bd82) [inline src/actions/actionsFactory.ts:13]: The action factory is one of the main flows that should be tested (manually)
As this is one of the only runtime changes to the production code

File: src/actions/impl.ts [hunk 35]
```diff
@@ -1,57 +1,58 @@
...
 
     constructor() {
-        super();
-        this.actionType = ActionType.Snippet;
+        this.actionType = SNIPPET;
         this.contributorId = "";
         this.snippetName = "";
-        this.context = "";
+        this.context = {};
+        this.isNonInteractive = false;
```
Comment (bd82) [inline src/actions/impl.ts:45]: I doubt this initialization to `isNonInteractive` would break anything. 
But maybe its redundant as we will be removing this class in the future anyhow
and the property itself is optional

File: tests/actionsFactory.spec.ts [hunk 36]
```diff
@@ -8,11 +8,10 @@ const testVscode = {
 
 mockVscode(testVscode, "src/actions/actionsFactory.ts");
 mockVscode(testVscode, "src/actions/impl.ts");
-mockVscode(testVscode, "src/actions/interfaces.ts");
 import { ActionsFactory } from "../src/actions/actionsFactory";
-import { CommandAction, ExecuteAction, FileAction, SnippetAction } from "../src/actions/impl";
-import { ActionType, ActionJsonKey } from '../src/actions/interfaces';
-
+import { ICommandAction, ISnippetAction } from '@sap-devx/app-studio-toolkit-types';
+import { ExecuteAction, } from "../src/actions/impl";
```
Comment (bd82) [inline tests/actionsFactory.spec.ts:13]: redundant comma

File: types/api.d.ts [hunk 37]
```diff
@@ -0,0 +1,121 @@
...
+    /**
+     * This an async wrapper for vscode extension that gets an extension by its full identifier and also wait until the target extension became active.
```
Comment (bd82) [inline types/api.d.ts:5]: its not a wrapper for vscode extension
its a wrapper for a vscode API.
Please link to the relevant API docs of vscode

File: types/api.d.ts [hunk 38]
```diff
@@ -0,0 +1,121 @@
+import {Uri} from 'vscode';
```
Comment (bd82) [inline types/api.d.ts:1]: I will go over the .d.ts and add a commit with docs upgrades

File: types/api.d.ts [hunk 39]
```diff
@@ -0,0 +1,155 @@
...
+   */
+  getParameter: (key: string) => Promise<string | undefined>;
+
+  /**
+   * @param key - URL parameter **key** name.
+   *
+   * @returns The **value** of the specified URL `key` IFF such a key exists in the BAS URL,
+   *          `undefined` otherwise.
+   */
+  getURLParamValue: (key: string) => Promise<string | undefined>;
```
Comment (bd82) [inline types/api.d.ts:45]: Ido says this is abstract on purpose, remove this newly named function

File: sample-action-client/package-lock.json [hunk 40]
```diff
@@ -1,25 +1,27 @@
...
-			"resolved": "https://registry.npmjs.org/@types/vscode/-/vscode-1.55.0.tgz",
-			"integrity": "sha512-49hysH7jneTQoSC8TWbAi7nKK9Lc5osQNjmDHVosrcU8o3jecD9GrK0Qyul8q4aGPSXRfNGqIp9CBdb13akETg=="
+			"resolved": "http://nexus.wdf.sap.corp:8081/nexus/repository/build.milestones.npm/@types/vscode/-/vscode-1.55.0.tgz",
+			"integrity": "sha512-49hysH7jneTQoSC8TWbAi7nKK9Lc5osQNjmDHVosrcU8o3jecD9GrK0Qyul8q4aGPSXRfNGqIp9CBdb13akETg==",
+			"dev": true
 		},
 		"ansi-styles": {
 			"version": "3.2.1",
-			"resolved": "https://registry.npmjs.org/ansi-styles/-/ansi-styles-3.2.1.tgz",
+			"resolved": "http://nexus.wdf.sap.corp:8081/nexus/repository/build.milestones.npm/ansi-styles/-/ansi-styles-3.2.1.tgz",
```
Comment (bd82) [inline sample-action-client/package-lock.json:24]: there should never be references to wdf.sap.corp on a SAP OSS

File: sample-action-client/package.json [hunk 41]
```diff
@@ -1,7 +1,7 @@
 {
 	"name": "sample-action-client",
 	"displayName": "Sample Action Client",
-	"version": "0.0.5",
+	"version": "0.0.6",
```
Comment (bd82) [inline sample-action-client/package.json:4]: I doubt this version really matters as this should not be released anywhere...

**Review by NetaMolcho — APPROVED** [review]
